core/cas-server-core-services/src/main/java/org/apereo/cas/services/DomainServicesManager.java

```diff
@@ -0,0 +1,197 @@
+package org.apereo.cas.services;
+
+import org.apereo.cas.authentication.principal.Service;
+import org.apereo.cas.support.events.service.CasRegisteredServiceDeletedEvent;
+import org.apereo.cas.support.events.service.CasRegisteredServiceSavedEvent;
+import org.apereo.inspektr.audit.annotation.Audit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.ApplicationEvent;
+import org.springframework.context.ApplicationEventPublisher;
+import org.springframework.scheduling.annotation.Scheduled;
+
+import javax.annotation.PostConstruct;
+import java.io.Serializable;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/**
+ * Default implementation of the {@link ServicesManager} interface.
+ *
+ * @author Scott Battaglia
+ * @since 3.1
+ */
+public class DomainServicesManager implements ServicesManager, Serializable {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultServicesManager.class);
+    private static final long serialVersionUID = -8581398063126547772L;
+
+    private final ServiceRegistryDao serviceRegistryDao;
+
+    @Autowired
+    private transient ApplicationEventPublisher eventPublisher;
+
+    private Map<Long, RegisteredService> services = new ConcurrentHashMap<>();
+    private Map<String, TreeSet<RegisteredService>> domains = new ConcurrentHashMap<>();
+
+    Pattern domainPattern = Pattern.compile("^https?://([^:/]+)");
+
+    /**
+     * Instantiates a new default services manager impl.
+     *
+     * @param serviceRegistryDao the service registry dao
+     */
+    public DomainServicesManager(final ServiceRegistryDao serviceRegistryDao) {
+        this.serviceRegistryDao = serviceRegistryDao;
+    }
+
+    @Audit(action = "DELETE_SERVICE", actionResolverName = "DELETE_SERVICE_ACTION_RESOLVER",
+            resourceResolverName = "DELETE_SERVICE_RESOURCE_RESOLVER")
+    @Override
+    public synchronized RegisteredService delete(final long id) {
+        final RegisteredService service = findServiceBy(id);
+        if (service != null) {
+            this.serviceRegistryDao.delete(service);
+            this.services.remove(id);
+            this.domains.get(getDomain(service.getServiceId())).remove(service);
+            publishEvent(new CasRegisteredServiceDeletedEvent(this, service));
+        }
+        return service;
+    }
+
+    @Override
+    public RegisteredService findServiceBy(final Service service) {
+        return service != null ? findServiceBy(service.getId()) : null;
+    }
+
+    @Override
+    public Collection<RegisteredService> findServiceBy(final Predicate<RegisteredService> predicate) {
+        return services.values().stream()
+                .filter(predicate)
+                .sorted()
```

Comment:
This is needed because the services are no longer stored in an ordered collection.  They are stored only in HashMaps, so the sorted is needed to ensure evaluation order is kept, especially in the current CAS Management app that still uses single list to present services and not by domain.